### PR TITLE
Adding build coverage cloud function, and refactoring test_utils.py

### PR DESCRIPTION
**Pull request by kabeer27**
-Added func to deploy script and main.py
-Added another argument to deploy script --> ./deploy.sh <project-name> <base-project-name> to be able to deploy the base image builder in the correct project
-Added unit test for request_coverage_build
-Tested request_coverage_build steps on cloud function
-Could not test coverage build due to my test-project service account not having access to download corpora from clusterfuzz.
-Renamed sync.py to project_sync.py to avoid import cache conflict with infra/gcb/sync.py
-Fixed sys.exit() to raise Runtime error when project name is missing from payload

@oliverchang 
@Dor1s 
@inferno-chromium

### CODE REVIEW

File: infra/build/functions/request_build.py [hunk 1]
```diff
@@ -85,3 +75,19 @@ def request_build(event, context):
...
+
+
+# pylint: disable=no-member
+def request_build(event, context):
+  """Entry point for cloud function to request builds."""
+  del context  #unused
+  if 'data' in event:
+    project_name = base64.b64decode(event['data']).decode('utf-8')
+  else:
+    logging.error('Project name missing from payload')
```
Comment (oliverchang) [inline infra/build/functions/request_build.py:87]: While you're here, fix this up to raise RuntimeError instead.

File: infra/build/functions/request_coverage_build.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                                                       base_images_project)
+  return build_steps
+
+
+def request_coverage_build(event, context):
+  """Entry point for coverage build cloud function."""
+  del event, context  #unused
+
+  with ndb.Client().context():
+
```
Comment (oliverchang) [inline infra/build/functions/request_coverage_build.py:46]: remove unnecessary newline.

File: infra/build/functions/request_coverage_build.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    credentials, image_project = google.auth.default()
+    for project in Project.query():
+      project_name = project.name
+      project_yaml_contents = project.project_yaml_contents
+      dockerfile_content = project.dockerfile_content
+      dockerfile_lines = dockerfile_content.split('\n')
```
Comment (oliverchang) [inline infra/build/functions/request_coverage_build.py:52]: Just initialize as `dockerfile_lines = project.dockerfile_content.split('\n')` without the intermediate variable.

File: infra/build/functions/request_coverage_build_test.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+              return_value=[{
+                  'url': 'test_download'
+              }])
+  @mock.patch('datetime.datetime')
+  def test_get_coverage_build_steps(self, mocked_url, mocked_corpora_steps,
+                                    mocked_time):
+    """Test for get_build_steps."""
+    del mocked_url, mocked_corpora_steps, mocked_time
+    datetime.datetime = test_utils.SpoofedDatetime
+    project_yaml_contents = 'language: c++\nsanitizers:\n  - address\narchitectures:\n  - x86_64\n'
```
Comment (oliverchang) [inline infra/build/functions/request_coverage_build_test.py:54]: nit: break this up into multiple lines.

i.e.

```python
project_yaml_contents = (
   'language: c++\n'
   'sanitizers:\n'
   '  - address\n'
   ...)

File: infra/build/functions/request_coverage_build_test.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    dockerfile_lines = dockerfile_contents.split('\n')
+    build_steps = get_coverage_build_steps('test-project',
+                                           project_yaml_contents,
+                                           dockerfile_lines, image_project,
+                                           base_images_project)
+    self.assertEqual(build_steps, expected_coverage_build_steps)
+
+  @classmethod
+  def tearDownClass(cls):
+    # TODO: replace this with a cleaner way of killing the process
```
Comment (oliverchang) [inline infra/build/functions/request_coverage_build_test.py:77]: no need for this comment. comment in cleanup_emulator should be sufficient.

File: infra/build/functions/deploy.sh [hunk 6]
```diff
@@ -94,19 +102,30 @@ deploy_scheduler $BASE_IMAGE_SCHEDULER_JOB \
...
+
+deploy_scheduler $COVERAGE_BUILD_SCHEDULER_JOB \
+				 "$COVERAGE_BUILD_SCHEDULE" \
+				 $COVERAGE_BUILD_JOB_TOPIC \
+				 "$COVERAGE_BUILD_MESSAGE" \
+				 $PROJECT_ID
 
 deploy_cloud_function sync \
-					  project_sync \
+					  sync \
```
Comment (Dor1s) [inline infra/build/functions/deploy.sh:114]: shouldn't this be `project_sync` as you renamed `sync.py` -> `project_sync.py`?

File: infra/build/functions/request_build_test.py [hunk 7]
```diff
@@ -71,9 +75,8 @@ def test_get_build_steps_no_project(self):
 
   @classmethod
   def tearDownClass(cls):
-    # TODO: replace this with a cleaner way of killing the process
```
Comment (Dor1s) [inline infra/build/functions/request_build_test.py:88]: was this TODO addressed or why are we removing it?

File: infra/build/functions/deploy.sh (hunk 6 above)
Comment (kabeer27) [reply infra/build/functions/deploy.sh:114]: This is just the display name that will be shown on the cloud platform, I'll chang it if required

File: infra/build/functions/request_build_test.py (hunk 7 above)
Comment (kabeer27) [reply infra/build/functions/request_build_test.py:88]: This was common code so it has been moved to test_utils.py the comment is there in that file

File: infra/build/functions/build_and_run_coverage.py [hunk 8]
```diff
@@ -0,0 +1 @@
+../../gcb/build_and_run_coverage.py
```
Comment (Dor1s) [inline infra/build/functions/build_and_run_coverage.py:1]: this is a symlink?

File: infra/build/functions/request_coverage_build.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+from google.cloud import ndb
+
+import build_and_run_coverage
+from datastore_entities import Project
+import request_build
+
+BASE_PROJECT = 'oss-fuzz-base'
+
+
+def get_coverage_build_steps(project_name, project_yaml_contents,
```
Comment (Dor1s) [inline infra/build/functions/request_coverage_build.py:28]: is this function really needed? looks like it simply wraps the other function which can be called directly

File: infra/build/functions/request_coverage_build.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+                                                       project_yaml_contents,
+                                                       dockerfile_lines,
+                                                       image_project,
+                                                       base_images_project)
+  return build_steps
+
+
+def request_coverage_build(event, context):
+  """Entry point for coverage build cloud function."""
+  del event, context  #unused
```
Comment (Dor1s) [inline infra/build/functions/request_coverage_build.py:42]: is `#unused` needed?

File: infra/build/functions/build_and_run_coverage.py (hunk 8 above)
Comment (kabeer27) [reply infra/build/functions/build_and_run_coverage.py:1]: Yes

**Comment by kabeer27** [comment]
Before it is merged i have one concern,
Currently since i am reusing the build_and_run_coverage code in a single cloud function to build for all projects,
The current code was made for a single build at a time, hence when a project is disabled or for any other reason skip_build() is called it goes for sys.exit(), so if a sys.exit() is called all the remaining coverage builds will not be built as well.

Proposed solution:
I could make a new_skip_build() method in my own module, then override the existing one => build_and_run_coverage.skip_build = new_skip_build

Let me know if this works @oliverchang @Dor1s

File: infra/build/functions/request_coverage_build.py (hunk 9 above)
Comment (oliverchang) [reply infra/build/functions/request_coverage_build.py:28]: +1

File: infra/build/functions/request_coverage_build.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    for project in Project.query():
+      project_name = project.name
+      project_yaml_contents = project.project_yaml_contents
+      dockerfile_lines = project.dockerfile_contents.split('\n')
+      # Todo, remove sys.exit call after infra migration
+      try:
+        build_steps = build_and_run_coverage.get_build_steps(
+            project_name, project_yaml_contents, dockerfile_lines,
+            image_project, BASE_PROJECT)
+      except SystemExit:
```
Comment (oliverchang) [inline infra/build/functions/request_coverage_build.py:43]: Actually, please add a comment here.

File: infra/build/functions/request_coverage_build.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  """Entry point for coverage build cloud function."""
+  del event, context  #unused
+
+  with ndb.Client().context():
+    credentials, image_project = google.auth.default()
+    for project in Project.query():
+      project_name = project.name
+      project_yaml_contents = project.project_yaml_contents
+      dockerfile_lines = project.dockerfile_contents.split('\n')
+      # Todo, remove sys.exit call after infra migration
```
Comment (oliverchang) [inline infra/build/functions/request_coverage_build.py:38]: Format as TODO: remove sys.exit call after infra migration.

File: infra/build/functions/request_coverage_build.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  """Entry point for coverage build cloud function."""
+  del event, context  #unused
+
+  with ndb.Client().context():
+    credentials, image_project = google.auth.default()
+    for project in Project.query():
+      project_name = project.name
+      project_yaml_contents = project.project_yaml_contents
+      dockerfile_lines = project.dockerfile_contents.split('\n')
+      # Todo: remove sys.exit call after infra migration
```
Comment (oliverchang) [inline infra/build/functions/request_coverage_build.py:38]: sorry missed that this was in the wrong file. this should be in build_and_run_coverage.py, where the call is actually being made.

Comment (kabeer27) [reply infra/build/functions/request_coverage_build.py:38]: Alright

**Pull request merged**